### ISSUE TITLE
Gradient Descent Clarification

### ISSUE THREAD

**Issue by gregynard** [The-Independent-Code/Neural-Network#3]
Hello, I'm a little bit confused by the gradient descent used on the neural network. Is it vanilla gradient descent or stochastic gradient descent? Correct me if I'm wrong but based on the code you update the weights right away after every input which is like a stochastic but you did not randomly select an input for each epoch instead you iterate all of them which is the part where I got confused.

**Comment by omaraflak** [comment]
Hi, yes this is stochastic gradient descent, since I'm updating after each input sample. It's true that training data should ideally be shuffled before looping through it.

As a clarification, I don't think stochastic gradient descent only goes through one randomly selected sample at each loop, as you seem to suggest. I believe you have to go through all the samples, but shuffle them before.

**Comment by gregynard** [comment]
Last question would I be able to use the same code and modify it as  vanilla, I hope this is not too much but would you be able to give me an idea or even basic guide on how to change it to vanilla gradient? Btw, thank you so much for your tutorials from hundreds of guides I've read and watched you are the one that has the simplest yet elegant and modular code.

**Comment by omaraflak** [comment]
No worries. You should be able to change it to regular gradient descent but it might make the code less clean :) I think the simplest way would be to sum the gradients in the backward() method, and move the weights/biases update logic to a new method, and call that method in the training loop after going through all samples.

**Comment by gregynard** [comment]
Thank you!
